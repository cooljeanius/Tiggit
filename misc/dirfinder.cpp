#include "dirfinder.hpp"

#include <boost/filesystem.hpp>
#include <stdlib.h>
#include <fstream>

using namespace Misc;
using namespace boost::filesystem;

#ifdef _WIN32

#include <windows.h>
#include <shlobj.h>
#include <tchar.h>

TCHAR pathbuf[MAX_PATH];

static std::string getPathCSIDL(int csidl)
{
  SHGetFolderPath(NULL, csidl, NULL, 0, pathbuf);
  return std::string(pathbuf);
}

std::string DirFinder::getAppData()
{
  return getPathCSIDL(CSIDL_LOCAL_APPDATA);
}

std::string DirFinder::getExePath()
{
  char buf[2000];
  GetModuleFileName(NULL, buf, 2000);
  return std::string(buf);
}

static std::string getDataDir(int csidl, const std::string &vname, const std::string &aname)
{
  path p = getPathCSIDL(csidl);
  if(vname != "") p /= vname;
  return (p / aname).string();
}

bool DirFinder::getStandardPath(std::string &dir)
{
  dir = getDataDir(CSIDL_LOCAL_APPDATA, vname, aname);
  return isWritable(dir);
}

static bool openRegKey(const std::string &vname, const std::string &aname, HKEY &hkey)
{
  std::string where = "Software\\" + vname + "\\" + aname;
  LPCTSTR sk = where.c_str();

  LONG openRes = RegCreateKeyEx(HKEY_CURRENT_USER,
                                sk,
                                0,NULL,
				REG_OPTION_NON_VOLATILE,
				KEY_ALL_ACCESS,NULL,
                                &hkey,
                                NULL);

  return openRes == ERROR_SUCCESS;
}


static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return false;

  DWORD dwSize = dir.size();
  if(RegSetValueEx(hkey, dname.c_str(), 0, REG_SZ,
                   (const BYTE*)dir.c_str(), dwSize) != ERROR_SUCCESS)
    return false;

  RegCloseKey(hkey);
  return true;
}

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return "";

  DWORD dwType = REG_SZ;
  DWORD dwSize = MAX_PATH;
  if(RegQueryValueEx(hkey, dname.c_str(), 0, &dwType, 
                     (BYTE*)pathbuf, &dwSize) != ERROR_SUCCESS)
    return "";

  RegCloseKey(hkey);

  return std::string(pathbuf);
}

#else
#ifdef __unix

static std::string getHome(const std::string &aname)
{
  /* More and more systems and apps are using ~/.local/share/appname
     instead of ~/.appname now. And I think that's a much less messy
     solution.
   */
  return std::string(getenv("HOME")) + "/.local/share/" + aname + "/";
}

static std::string getPathFile(const std::string &aname,
                               const std::string &dname)
{
  return getHome(aname) + dname + ".conf";
}

std::string DirFinder::getAppData() { return ""; }

bool DirFinder::getStandardPath(std::string &dir)
{
  dir = getHome(aname);
  return isWritable(dir);
}

std::string DirFinder::getExePath() { return ""; }

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  std::string file = getPathFile(aname,dname);
  if(exists(file))
    {
      std::ifstream inp(file.c_str());
      if(inp)
        {
          std::string res;
          inp >> res;
          if(inp)
            return res;
        }
    }
  return "";
}

static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  std::string file = getPathFile(aname,dname);
  create_directories(path(file).parent_path());
  std::ofstream out(file.c_str());
  if(out)
    {
      out << dir;
      out.flush();
      if(out)
        return true;
    }
  return false;
}

#else
#error "This platform is not supported yet."
#endif
#endif

bool DirFinder::getStoredPath(std::string &dir)
{
  dir = os_getStoredPath(vname,aname,dname);
  return isWritable(dir);
}

bool DirFinder::setStoredPath(const std::string &dir)
{
  return isWritable(dir) && os_setStoredPath(dir,vname,aname,dname);
}

bool DirFinder::isWritable(const std::string &dir)
{
  if(dir == "")
    return false;

  try
    {
      // Create directories
      create_directories(dir);

      // Do write check
      path p = dir;
      p /= "xyz.tmp";
      std::string file = p.string();
      {
        std::ofstream out(file.c_str());
        if(!out) return false;
        out << "abc";
        if(!out) return false;
      }
      {
        std::ifstream inp(file.c_str());
        if(!inp) return false;
        std::string tmp;
        inp >> tmp;
        if(!inp || tmp != "abc")
          return false;
      }

      remove(p);
    }
  // All errors means the dir is not writable
  catch(...) { return false; }
  return true;
}
