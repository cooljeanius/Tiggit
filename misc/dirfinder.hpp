#ifndef __DIRFINDER_HPP_
#define __DIRFINDER_HPP_

#include <string>

/*
  This class finds an appropriate writable data path for the
  application.

  It can propose platform-dependent standard paths, or accept a
  user-provided value. It will remember settings between sessions.

  We only accept directories that are writable, and we test dirs for
  writability.

  You can have multiple data directories and store independent
  settings for each, if you provide different names to the
  constructor. The default name is 'default'.

  NOTE: On Windows Vista and newer, UAC virtualization will render
  write tests useless for determining access rights. It's highly
  recommended to disable virtualization by including a manifest file
  with your application, and specifing your privileges as "asInvoker"
  or similar.
  More info:
      http://msdn.microsoft.com/en-us/library/bb756960.aspx
      http://msdn.microsoft.com/en-us/library/bb756929.aspx

  Usage:

  A good practice is to call getStoredPath() first, and ALWAYS prompt
  the user if it returns false. The path returned by getStandardPath()
  may then be presented to the user as a default value.

  Whatever the user supplies, return it to DirFinder through
  setStoredPath().
 */

namespace Misc
{
  struct DirFinder
  {
    /* The vendor_name, app_name and dirname should be filesystem and
       windows registry compatible names. Preferably short, lower-case
       and without spaces.
     */
    DirFinder(const std::string &vendor_name,
              const std::string &app_name,
              const std::string &dirname = "default")
      : vname(vendor_name), aname(app_name), dname("path-" + dirname) {}

    /* Get stored path. Returns true if found and writable, false
       otherwise.

       If a path is set but the return value is false, it means the
       stored path was invalid or not writable.
     */
    bool getStoredPath(std::string &path);

    /* Get OS-dependent standard path for applications. Will only
       return true if an appropriate writable path was found.

       This does NOT change the stored path, but it may create the
       given path if it does not exist. Use setStoredPath() to use
       this path.
     */
    bool getStandardPath(std::string &path);

    /* Windows only. Returns CSIDL_LOCAL_APPDATA folder. Returns an
       empty string on other platforms.
     */
    static std::string getAppData();

    /* Get current executable path. Currently windows only.
     */
    static std::string getExePath();

    /* Set stored path. Will return true if the path was accepted,
       ie. it was writable, and we were able to store the information
       in a retrievable way.

       If the path was accepted, future requests to getStoredPath()
       will return this path.
     */
    bool setStoredPath(const std::string &path);

    /* Tests if a given directory is valid and writable as a data
       path. Will create directories if they don't exist.
     */
    static bool isWritable(const std::string &path);

  private:
    std::string vname,aname,dname;
  };
}

#endif
