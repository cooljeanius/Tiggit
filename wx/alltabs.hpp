/* This is a direct include file for frame.cpp. It is not meant to be
   included anywhere else.
 */
struct TitleCol : ColumnHandler
{
  bool addStatus;

  TitleCol(bool as=false) : addStatus(as) {}

  bool doSort(wxGameList &lst) { return lst.sortTitle(); }
  wxString getText(const wxGameInfo &e)
  {
    return e.getTitle(addStatus);
  }
};

struct AddDateCol : ColumnHandler
{
  bool doSort(wxGameList &lst) { return lst.sortDate(); }
  wxString getText(const wxGameInfo &e)
  {
    return e.timeString();
  }
};

struct TypeCol : ColumnHandler
{
  wxString demoText, freeText;

  TypeCol()
  {
    demoText = wxT("demo");
    freeText = wxT("free");
  }

  bool doSort(wxGameList &lst) { return false; }
  wxString getText(const wxGameInfo &e)
  {
    if(e.isDemo())
      return demoText;
    return freeText;
  }
};

struct RatingCol : ColumnHandler
{
  bool doSort(wxGameList &lst) { return lst.sortRating(); }
  wxString getText(const wxGameInfo &e)
  {
    return e.rateString();
  }
};

struct DownloadsCol : ColumnHandler
{
  bool doSort(wxGameList &lst) { return lst.sortDownloads(); }
  wxString getText(const wxGameInfo &e)
  {
    return e.dlString();
  }
};

struct StatusCol : ColumnHandler
{
  wxString textNotInst, textReady;

  bool doSort(wxGameList &lst) { return false; }
  StatusCol()
  {
    textNotInst = wxT("Not installed");
    textReady = wxT("Ready to play");
  }

  wxString getText(const wxGameInfo &e)
  {
    if(e.isUninstalled())
      return textNotInst;

    if(e.isInstalled())
      return textReady;

    return e.statusString();
  }
};

/*
struct AllGamesTab : GameTab
{
  AllGamesTab(wxNotebook *parent, wxGameData &data)
    : GameTab(parent, wxT("All Games"), data.getAllList(), data)
  {
    list->addColumn("Name", 310, new TitleCol);
    list->addColumn("Status", 170, new StatusCol);
  }
};
*/

// Lists newly added games
struct NewGamesTab : GameTab
{
  NewGamesTab(wxNotebook *parent, wxGameData &data)
    : GameTab(parent, wxT("Latest"), data.getLatest(), data)
  {
    list->addColumn("Name", 370, new TitleCol(true));
    list->addColumn("Added", 120, new AddDateCol);
    list->markNew = true;

    lister.sortDate();
  }

  int getTitleNumber()
  {
    int newGames = 0;

    // Count number of new games
    for(int i=0; i<lister.size(); i++)
      if(lister.get(i).isNew())
        newGames++;

    return newGames;
  }
};

// Lists freeware games
struct FreewareTab : GameTab
{
  FreewareTab(wxNotebook *parent, wxGameData &data)
    : GameTab(parent, wxT("Freeware"), data.getFreeware(), data)
  {
    list->addColumn("Name", 310, new TitleCol(true));
    list->addColumn("Rating", 95, new RatingCol);
    list->addColumn("Downloads", 75, new DownloadsCol);

    list->markInstalled = true;

    lister.sortRating();
  }
};

// Lists demos
struct DemoTab : GameTab
{
  DemoTab(wxNotebook *parent, wxGameData &data)
    : GameTab(parent, wxT("Demos"), data.getDemos(), data)
  {
    list->addColumn("Name", 310, new TitleCol(true));
    list->addColumn("Rating", 95, new RatingCol);
    list->addColumn("Downloads", 75, new DownloadsCol);

    list->markInstalled = true;

    lister.sortRating();
  }
};

// Installed or installing games
struct InstalledTab : GameTab
{
  InstalledTab(wxNotebook *parent, wxGameData &data)
    : GameTab(parent, wxT("Installed"), data.getInstalled(), data)
  {
    list->addColumn("Name", 310, new TitleCol);
    list->addColumn("Status", 170, new StatusCol);

    lister.sortTitle();
  }
};
