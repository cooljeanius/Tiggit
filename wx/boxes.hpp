#ifndef __WX_BOXES_HPP_
#define __WX_BOXES_HPP_

#include "wxcommon.hpp"

namespace wxTiggit
{
  struct Boxes
  {
    // Ask the user an OK/Cancel question.
    static bool ask(const wxString &question)
    {
      return wxMessageBox(question, wxT("Please confirm"),
                          wxOK | wxCANCEL | wxICON_QUESTION) == wxOK;
    }

    // Display an error message box
    static void error(const wxString &msg)
    {
      wxMessageBox(msg, wxT("Error"), wxOK | wxICON_ERROR);
    }

    static void say(const wxString &msg)
    {
      wxMessageBox(msg, wxT("Information"), wxOK | wxICON_INFORMATION);
    }

    static bool ask(const std::string &q) { return ask(strToWx(q)); }
    static void error(const std::string &q) { error(strToWx(q)); }
    static void say(const std::string &q) { say(strToWx(q)); }
  };
}
#endif
