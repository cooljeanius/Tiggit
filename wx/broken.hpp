/* This is a drop-in include for gametab.cpp, not meant to be used
   separately.
 */
struct ProblemDialog : wxDialog
{
  bool ok;

  std::string comment;

  ProblemDialog(wxWindow *parent, const wxString &gameTitle)
    : wxDialog(parent, -1, wxT("Report Broken Download"), wxDefaultPosition,
               wxSize(300,300))
  {
    wxPanel *panel;
    wxBoxSizer *sizer;

    panel = new wxPanel(this);

    // Set up the sizers
    wxBoxSizer *outer = new wxBoxSizer(wxVERTICAL);
    sizer = new wxBoxSizer(wxVERTICAL);
    panel->SetSizer(outer);
    outer->Add(sizer, 1, wxGROW | wxALL, 20);

    wxTextCtrl *t_comment;

    sizer->Add(new wxStaticText(panel, -1, wxT("Game: ") + gameTitle));
    sizer->Add(new wxStaticText(panel, -1, wxT("Describe the issue (required):")), 0, wxTOP, 6);
    t_comment = new wxTextCtrl(panel, -1, wxT(""), wxDefaultPosition, wxSize(260,80),
                               wxTE_MULTILINE);
    sizer->Add(t_comment);

    wxBoxSizer *buttons = new wxBoxSizer(wxHORIZONTAL);
    sizer->Add(buttons, 0, wxTOP, 12);
    buttons->Add(new wxButton(panel, wxID_OK, wxT("OK")));
    buttons->Add(new wxButton(panel, wxID_CANCEL, wxT("Cancel")));

    panel->Fit();
    Fit();
    Center();

    ok = ShowModal() == wxID_OK;

    if(ok)
      {
        comment = wxToStr(t_comment->GetValue());

        // Not filling out the description is considered the same as
        // clicking cancel
        if(comment == "") ok = false;
      }

    Destroy();
  }
};
