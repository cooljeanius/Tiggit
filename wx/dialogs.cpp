#include "dialogs.hpp"

using namespace wxTiggit;

#define myID_BROWSE 12345

BrowseDialog::BrowseDialog(wxWindow *parent, const wxString &title,
                           int width, int height, bool file)
  : wxDialog(parent, -1, title, wxDefaultPosition, wxSize(width, height)),
    browseFile(file)
{
  Connect(myID_BROWSE, wxEVT_COMMAND_BUTTON_CLICKED,
          wxCommandEventHandler(BrowseDialog::onBrowse));
}

void BrowseDialog::addTo(wxSizer *sizer, const wxString &value)
{
  wxBoxSizer *hbox = new wxBoxSizer(wxHORIZONTAL);
  edit = new wxTextCtrl(this, -1, value, wxDefaultPosition, wxSize(300,24));
  hbox->Add(edit, 0, wxTOP, 2);
  hbox->Add(new wxButton(this, myID_BROWSE, wxT("Browse")), 0, wxLEFT, 5);
  sizer->Add(hbox);
}

void BrowseDialog::onBrowse(wxCommandEvent &event)
{
  wxString dir = edit->GetValue();
  if(browseFile)
    dir = wxFileSelector(wxT("Chose file"), dir);
  else
    dir = wxDirSelector(wxT("Choose directory"), dir);
  if(!dir.empty())
    edit->ChangeValue(dir);
}

OutputDirDialog::OutputDirDialog(wxWindow *parent, const std::string &old_dir,
                                 const std::string &legacy_dir,
                                 bool writeFailed, bool freshInstall)
  : BrowseDialog(parent, wxT("Select Data Directory"), 430, 256)
{
  wxString old(old_dir.c_str(), wxConvUTF8);

  wxBoxSizer *vbox = new wxBoxSizer(wxVERTICAL);

  wxString text = wxT("Select a data directory. This is where all games, configuration and data\nwill be stored.");

  vbox->Add(new wxStaticText(this, -1, text), 0, wxBOTTOM, 15);

  vbox->Add(new wxStaticText(this, -1, wxT("Data directory:")), 0, wxBOTTOM, 3);
  addTo(vbox,old);
  if(writeFailed)
    {
      wxStaticText *t = new wxStaticText(this, -1, wxT("The path you specified is not writable!"));
      t->SetForegroundColour(wxColour(255,0,0));
      vbox->Add(t);
    }

  if(freshInstall)
    text = wxT("The directory will be created if it does not already exist.\nPress Cancel to exit.");
  else
    text = wxT("WARNING: This will MOVE all installed games, savegames and\nTiggit configuration to this location.");
  vbox->Add(new wxStaticText(this, -1, text),0, wxTOP, 12);

  wxCheckBox *import = NULL;
  if(legacy_dir != "")
    {
      assert(freshInstall);

      import = new wxCheckBox(this, -1, strToWx("Import games from " + legacy_dir + " \n(files are moved to the new directory)"));
      import->SetValue(true);
      vbox->Add(import, 0, wxTOP, 10);
    }

  vbox->Add(CreateButtonSizer(wxOK | wxCANCEL), 0, wxALIGN_CENTER | wxTOP, 20);

  wxBoxSizer *main = new wxBoxSizer(wxVERTICAL);
  main->Add(vbox, 0, wxALL, 15);
  SetSizer(main);

  // This dialog is kinda dynamic in content, so better to ask it to
  // resize itself depending on contained elements.
  Fit();

  Centre();

  // Run and fetch info
  ok = ShowModal() == wxID_OK;
  //move = moverb->GetValue();

  doImport = false;
  if(legacy_dir != "" && import)
    doImport = import->GetValue();

  move = true;
  path = std::string(edit->GetValue().mb_str());
  changed = old_dir != path;

  Destroy(); 
}
/*
ImportDialog::ImportDialog(wxWindow *parent, const std::string &maindir)
  : BrowseDialog(parent, wxT("Add/Import Data"), 450, 280)
{
  wxBoxSizer *vbox = new wxBoxSizer(wxVERTICAL);

  wxString old(maindir.c_str(), wxConvUTF8);

  vbox->Add(new wxStaticText(this, -1, wxT("Select a directory that was previously created through\nthe Export option.")),
            0, wxBOTTOM, 30);

  vbox->Add(new wxStaticText(this, -1, wxT("Directory to import:")));
  addTo(vbox);

  // Not really needed, and not sure how to best implement it.
  //wxStaticText *status = new wxStaticText(this, -1, wxT("Found: xyz/installed.json - OK"));
  //status->SetForegroundColour(wxColour(0,128,0));
  //vbox->Add(status, 0, wxLEFT, 3);

  wxRadioButton *rb = new wxRadioButton(this, -1, wxT("Play from this location"),
                                        wxDefaultPosition, wxDefaultSize, wxRB_GROUP);
  vbox->Add(rb, 0, wxTOP, 25);
  vbox->Add(new wxRadioButton(this, -1, wxT("Copy files to ") + old));
  vbox->Add(CreateButtonSizer(wxOK | wxCANCEL), 0, wxALIGN_CENTER | wxTOP, 30);

  wxBoxSizer *main = new wxBoxSizer(wxVERTICAL);
  main->Add(vbox, 0, wxALL, 8);
  SetSizer(main);

  Centre();

  // Run and fetch info
  ok = ShowModal() == wxID_OK;
  copy = !rb->GetValue();

  source = std::string(edit->GetValue().mb_str());

  Destroy();
}

ExportDialog::ExportDialog(wxWindow *parent, const std::vector<std::string> &games)
  : BrowseDialog(parent, wxT("Export Data"), 450, 280)
{
  wxBoxSizer *vbox = new wxBoxSizer(wxVERTICAL);

  vbox->Add(new wxStaticText(this, -1, wxT("Select a directory where data should be exported. This is useful for\ntransmitting games through USB sticks and similar.\nExported data can be loaded with the Import menu option.")),
            0, wxBOTTOM, 30);

  vbox->Add(new wxStaticText(this, -1, wxT("Destination directory:")));
  addTo(vbox);

  wxRadioButton *rb = new wxRadioButton(this, -1, wxT("Include Tiggit launcher"),
                                            wxDefaultPosition, wxDefaultSize, wxRB_GROUP);
  vbox->Add(rb, 0, wxTOP, 25);
  vbox->Add(new wxRadioButton(this, -1, wxT("Copy games only")));
  vbox->Add(CreateButtonSizer(wxOK | wxCANCEL), 0, wxALIGN_CENTER | wxTOP, 30);

  wxBoxSizer *main = new wxBoxSizer(wxVERTICAL);
  main->Add(vbox, 0, wxALL, 8);
  SetSizer(main);

  Centre();

  // Run and fetch info
  ok = ShowModal() == wxID_OK;
  launcher = rb->GetValue();
  output = std::string(edit->GetValue().mb_str());

  Destroy();
}
*/
/* Checkbox list for the export dialog:
   http://wiki.wxwidgets.org/WxListCtrl#Implement_wxListCtrl_with_Checkboxes
   WebUpdate has a CheckedListCtrl which we'll steal.
 */
