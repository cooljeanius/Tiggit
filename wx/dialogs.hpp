#ifndef __TIGGIT_DIALOGS_HPP_
#define __TIGGIT_DIALOGS_HPP_

#include "wxcommon.hpp"
//#include <vector>

namespace wxTiggit
{
  // Base class for dialogs that have a "browse file/directory" entry.
  struct BrowseDialog : wxDialog
  {
    BrowseDialog(wxWindow *parent, const wxString &title,
                 int width, int height, bool file=false);

  protected:
    void addTo(wxSizer *sizer, const wxString &value=wxT(""));

    wxTextCtrl *edit;

  private:
    bool browseFile; // Find file or directory. Default is directory.
    void onBrowse(wxCommandEvent &event);
  };

  /* Select a new repository directory.

     Parameters:

       old_dir    - old directory (in case of moving), or suggestion for new
                    directory (in case of a fresh install)

       legacy_dir - If non-empty, prompt the user if they want to
                    import from the given directory. The answer is
                    recorded in doImport.

       writeFailed - display a message saying that the previously selected directory was not                      writable

       freshInstall - treat this as a fresh install, where the user gets to select the
                      directory for the first time

   */
  struct OutputDirDialog : BrowseDialog
  {
    OutputDirDialog(wxWindow *parent, const std::string &old_dir,
                    const std::string &legacy_dir = "",
                    bool writeFailed=false, bool freshInstall=false);

    bool ok, move, changed, doImport;
    std::string path;
  };

  /*
  struct ImportDialog : BrowseDialog
  {
    ImportDialog(wxWindow *parent, const std::string &maindir);

    bool ok, copy;
    std::string source;
  };

  struct ExportDialog : BrowseDialog
  {
    ExportDialog(wxWindow *parent, const std::vector<std::string> &games);

    bool ok, launcher;
    std::string output;
    std::vector<int> selected;
  };

  struct AddExternalDialog : BrowseDialog
  {
    AddExternalDialog(wxWindow *parent);

    bool ok;
    std::string name, exe;
  };
  */
}
#endif
