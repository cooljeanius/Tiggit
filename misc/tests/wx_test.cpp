#include "dirfinder.hpp"

#define wxUSE_UNICODE 1
#include <wx/stdpaths.h>
#include <wx/wx.h>

#include <iostream>
using namespace std;
using namespace Misc;

DirFinder fnd("tiggit.net", "tiggit", "finder-test");

void testWrite(const std::string &path)
{
  cout << path << ": ";
  if(DirFinder::isWritable(path))
    cout << "GOOD\n";
  else
    cout << "BAD\n";
}

class MyApp : public wxApp
{
public:
  virtual bool OnInit()
  {
    if(!wxApp::OnInit())
      return false;

    SetAppName(wxT("tiggit"));

    string wxdata(wxStandardPaths::Get().GetUserLocalDataDir().mb_str());
    cout << "wxW suggestion: ";
    testWrite(wxdata);

    cout << "Our suggestion: ";
    string out;
    fnd.getStandardPath(out);
    testWrite(out);
    return false;
  }
};

IMPLEMENT_APP(MyApp)
