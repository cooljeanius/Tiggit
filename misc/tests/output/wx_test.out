wxW suggestion: /home/mortennk/.tiggit: GOOD
Our suggestion: /home/mortennk/.local/share/tiggit/: GOOD
